### PR TITLE
Fix CakeSocket not being able to connect to TLS1.2 only servers

### PR DESCRIPTION
**Pull request by markstory**
Include the changes from #10232 and also fix the issue described in #10445

### CODE REVIEW

File: lib/Cake/Network/CakeSocket.php [hunk 1]
```diff
@@ -116,6 +117,32 @@ class CakeSocket {
  */
 	public function __construct($config = array()) {
 		$this->config = array_merge($this->_baseConfig, $config);
+
+		// These TLS versions are not supported by older PHP versions,
+		// so we have to conditionally set them if they are supported.
+		$conditionalCrypto = array(
+			'tlsv1_1_client' => 'STREAM_CRYPTO_METHOD_TLSv1_1_CLIENT',
```
Comment (ravage84) [inline lib/Cake/Network/CakeSocket.php]: If we set these TLS versions depending on the PHP version, shouldn't we test the list of them conditionally?

File: lib/Cake/Network/CakeSocket.php [hunk 2]
```diff
@@ -116,6 +117,32 @@ class CakeSocket {
  */
 	public function __construct($config = array()) {
 		$this->config = array_merge($this->_baseConfig, $config);
+
+		// These TLS versions are not supported by older PHP versions,
```
Comment (ravage84) [inline lib/Cake/Network/CakeSocket.php]: The whole TLS related code could be moved into a protected method in which the doc blcok could epxlain the situation.

 + smaller method
 + related stuff is extracted and contained
 + in line comment moves to class API  docs

File: lib/Cake/Network/CakeSocket.php [hunk 3]
```diff
@@ -44,7 +44,8 @@ class CakeSocket {
 		'host' => 'localhost',
 		'protocol' => 'tcp',
 		'port' => 80,
-		'timeout' => 30
+		'timeout' => 30,
+		'cryptoType' => 'tls'
```
Comment (ravage84) [inline lib/Cake/Network/CakeSocket.php]: Nitpicking: trailing comma.

**Review by ravage84 — COMMENTED** [review]
Looks good.

File: lib/Cake/Network/CakeSocket.php (hunk 1 above)
Comment (RichieB2B) [inline lib/Cake/Network/CakeSocket.php]: This is done by the `if (defined($const))` on line 130. If PHP doesn't know about the constant, it won't get added.

File: lib/Cake/Network/CakeSocket.php (hunk 2 above)
Comment (markstory) [inline lib/Cake/Network/CakeSocket.php]: good idea

File: lib/Cake/Network/CakeSocket.php (hunk 1 above)
Comment (espaistelematics) [reply lib/Cake/Network/CakeSocket.php]: Perfect! Works good.

**Pull request merged**